### ISSUE TITLE
class binding not updating on client after using `$isServer` with boolean values

### ISSUE THREAD

**Issue by AlbertMarashi** [vuejs/vue#9255]
### Version
2.5.21

### Reproduction link
[https://codesandbox.io/s/y05q554q71](https://codesandbox.io/s/y05q554q71)

### Steps to reproduce
code sandbox will illustrate the issue. (may take a few seconds to load)

Notice how `{{ isActive() }}` prints true, yet the class isn't binded

### What is expected?
`:class` to bind class attribute `class="nav-link active` and cause nav-link to turn red

### What is actually happening?
Not binding class and making nav-link red

**Comment by posva** [comment]
I would say this is intended, you don't want your component to completely re renders when hydrating, that would defeat the purpose of hydration. You can call `$forceUpdate` in `mounted` to make it happen. But the whole point of using `$isServer` is that you know when you rendered on server. And that's pretty much what you have right now

**Comment by DominusVilicus** [comment]
Thanks, @nickmessing provided the same solution over DMs, I was about to post it and close.

```js
<template>
    <a class="nav-link" :href="href" :class="{ active: isActive() }"><slot></slot></a> 
</template>
<script>
export default {
    props: ['href'],
    methods: {
        isActive(){
            if(this.$isServer) return false
            return window.location.pathname === this.href
        }
    },
    mounted(){
        this.$forceUpdate()
    }
}
</script>
<style scoped>
.nav-link {
  color: white;
  padding: 10px 14px;
  margin: 4px;
  background: blue;
}
.active {
  background: red;
}
</style>
```

```
OR:
```
<template>
  <a class="nav-link" v-bind:class="{ active: active }" :href="href">
    <slot></slot> {{ active }}</a
  >
</template>
<script>
export default {
  props: ["href"],
  data() {
    return { active: false };
  },
  mounted() {
    this.active = window.location.pathname === this.href;
  }
};
</script>
<style scoped>
.nav-link {
  color: white;
  padding: 10px 14px;
  margin: 4px;
  background: blue;
}
.active {
  background: red;
}
</style>
```

**Comment by lbogdan** [comment]
@posva I don't think that's entirely true, as the "Home" `<nav-link>` is actually re-rendered from showing "false", like it comes hydrated in the html, to showing "true" (because `isActive()` now returns `true` on the client). So what exactly is the rule for things to be re-hydrated or not?

**Comment by posva** [comment]
it's indeed weird, it only happens with `class`

**Comment by DominusVilicus** [comment]
Well, if I have a say, I would like to see class/element attributes hydrated (when `app.$mount("#el", true)` is passed, as that seems expected behaviour.

I spent a few hours unable to solve the issue, as there was no info in the docs about whether element attributes are hydrated or not.

To add, `isActive` was being called, whether it was a `computed` method or a `method`. If the class binding method is being ignored, then why call it at all?

**Comment by bny6613** [comment]
I have a similar issue. Here is what i found in the source code https://github.com/vuejs/vue/blob/dev/dist/vue.js#L6068
```
    // list of modules that can skip create hook during hydration because they
    // are already rendered on the client or has no need for initialization
    // Note: style is excluded because it relies on initial clone for future
    // deep updates (#7063).
    var isRenderedModule = makeMap('attrs,class,staticClass,staticStyle,key');
```
and then in `hydrate` function there is a check
```
if (!isRenderedModule(key)) {
    fullInvoke = true;
    invokeCreateHooks(vnode, insertedVnodeQueue);
    break
 }
```
So it turns out that Vue does not call invokeCreateHooks (which internally calls updateAttrs) during hydration if there are no other directives attached to the element. That is why your class attribute has not been updated. If you add any event or ref to that element, it will call the hooks.

**Comment by bny6613** [comment]
Removing `class` from `isRenderedModule` map fixes the issue, but i am not sure about performance implications and maybe there is a better way to fix it.

**Comment by DominusVilicus** [comment]
> So what exactly is the rule for things to be re-hydrated or not?
> it's indeed weird, it only happens with class

So why not just re-render everything, including class. It seems like intended behaviour. @yyx990803

**Comment by github-edu** [comment]
ok

2018-12-26 22:21 GMT+08:00, Eduardo San Martin Morote
<notifications@github.com>:

**Comment by github-edu** [comment]
no problem

2019-02-19 14:45 GMT+08:00, anymous google <imlaotse@gmail.com>:

**Comment by github-edu** [comment]
what is your mother

2019-02-19 14:45 GMT+08:00, anymous google <imlaotse@gmail.com>:

**Comment by DominusVilicus** [comment]
@yyx990803 can we fix this maybe in 2.6 or 3.0 if you're worried about it being a breaking change.
